### PR TITLE
[WIP] Build sap portal

### PR DESCRIPTION
**Pull request by polybuildr**
Work in progress, do not merge! But please review and leave comments. Meant to resolve T83.

### CODE REVIEW

**Comment by polybuildr** [comment]
Partially {resolves T83}.

File: src/app/views/sap/register.php [hunk 1]
```diff
@@ -9,14 +9,7 @@
...
 <body>
-    <div class="navbar collapsed" style="position:static;">
-        <a class="navbar-title" href="javascript:void(0)">Felicity ʼ16</a>
-        <ul class="pull-right">
-            <li><a href="<?= base_url() ?>">Home</a></li>
-        </ul>
-        <a href="javascript:void(0)" class="menu-toggle-button">&equiv;</a>
-        </a>
-    </div>
+    <?php $this->load_fragment('sap/navbar_fragment', ['logged_in' => $logged_in]); ?>
```
Comment (nisargjhaveri) [inline src/app/views/sap/register.php:12]: It already passes whatever is passed to the view. So, I don't think you need to pass `logged_in` again.

Comment (nisargjhaveri) [inline src/app/views/sap/register.php:12]: Well, it is more easier to comprehend this way. So, don't know if we should keep it or not.

Comment (polybuildr) [inline src/app/views/sap/register.php:12]: I didn't know that. I was wishing that the functionality was there, so I could use it. :P But you're right, it is easier to understand this way. Let's keep it like this until something changes.

**Comment by polybuildr** [comment]
Okay, I think I'm going to have to temporarily stop work on this now and hand it over to you guys, @nisargjhaveri, @hharchani. We need to get this deployed in the next 20 hours, during which time I won't be available for most of it.

A lot of things are not done, but these are the basics, and it should not be too difficult to build on top of this. Please merge this and implement on top of it, or of course, pull from my branch and edit and commit and new PR.

One major issue is that users don't get added to `sap_users` in any way, so you'll have to manually create users to test anything. (You can do the initial registration using the form, though.)

Things that work:
- Admins can create missions and create tasks within those missions. These tasks can also optionally include filling of a text field.
- Users can submit tasks, including filling the text field.
- Admins can review tasks,  "approve" or "reject"
- Users' view of a mission shows accepted tasks, also shows rejected tasks only first page load (sorry)
- Users can resubmit rejected tasks

Things that must be done:
- Awarding of scores. The `sap_portal` controller has a `TODO`. Once all tasks of a mission are completed, the user's score must increase by the amount specified in the mission. The score should also be displayed somewhere.
- A view where admins can view users who have registered, and "approve" them to join the sap (technically, copy users from `sap_ambassadors` to `sap_users`. Please look into removing the "username" field and replacing it with the email and additionally, the name or something. The username is currently only being used for logins.
- Database schema checking. At least one foreign key is missing. `registration_id` in `sap_user`, I think. Some `UNIQUE` checks are missing. While the UI does not support it, it is possible for an ambassador to make multiple task submissions at the moment.

Things that should be done:
- Leaderboard. Ranking of people by scores.
- Duplication of some fields across tables, with `ON UPDATE CASCADE`. Refer line 181 of `sap_model` in the current version of this PR.
- Lots more. Some maybe should go into the above category. :P I'm sorry, it's late, I'm tired and I'm also tired of working on this portal. Help!

**Comment by polybuildr** [comment]
More 'must's.
- `htmlspecialchars` some stuff that I'm printing without sanitizing

One thing to keep in mind is that you can trust SAP admins to not game the system, but not ambassadors - since we don't have any character checking of the ambassadors.

**Comment by polybuildr** [comment]
Attempted to fix the unsanitized output since we **really** don't want that.

**Comment by polybuildr** [comment]
Additional features that _must_ be there:
- Deleting a mission or a task

Additional features that _should_ be there:
- Editing/updating a mission or a task

File: src/app/controllers/sap.php [hunk 2]
```diff
@@ -34,10 +36,40 @@ function index() {
...
+                ]);
+            }
+        } else {
+            $this->load_view('sap/login');
+        }
+    }
+
+    public function logout() {
+        $this->load_library('sap_auth_lib', 'sap_auth');
+        $this->load_library('http_lib');
```
Comment (hharchani) [inline src/app/controllers/sap.php]: Why is this loaded?

Comment (polybuildr) [inline src/app/controllers/sap.php]: My bad, no reason.

File: src/app/models/sap_model.php [hunk 3]
```diff
@@ -22,4 +24,220 @@ public function registerEntry($data) {
...
+ON `tasks`.`id` = `task_submissions`.`task_id`
+INNER JOIN `sap_missions` as `missions`
+ON `missions`.`id` = `tasks`.`mission_id`
+INNER JOIN `sap_users` AS `users`
+ON `users`.`id` = `task_submissions`.`user_id`
+INNER JOIN `sap_ambassadors` AS `ambassadors`
+ON `ambassadors`.`id` = `users`.`registration_id`
+WHERE `tasks`.`mission_id`=? and `task_submissions`.`done` = 0
+SQL;
+        $stmt = $this->db_lib->prepared_execute(
```
Comment (hharchani) [inline src/app/models/sap_model.php:201]: Itni gandi query kyu hai? :( :disappointed:

Comment (polybuildr) [inline src/app/models/sap_model.php:201]: Look at line 181. :P

**Comment by polybuildr** [comment]
Rebased on master after #74 and also fixed a typo in the last commit message.

**Comment by hharchani** [comment]
Looks fine. Merging. @smritz will test it afterwards.

**Pull request merged**